File: rekall-model/src/main/java/dev/rekall/domain/RunQueue.java
package dev.rekall.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

/**
 * The one run queue: when it starts, the usage ceiling it stops at, and how the sessions it opens
 * are launched. There is a single row; the queued tasks are {@link RunQueueItem}s.
 */
@Entity
@Table(name = "run_queue")
@Getter
@Setter
public class RunQueue {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "id", nullable = false, updatable = false)
    private UUID id;

    @Enumerated(EnumType.STRING)
    @Column(name = "state", nullable = false, length = 20)
    private RunQueueState state = RunQueueState.IDLE;

    /** When a scheduled queue starts; null for one started at once or not started at all. */
    @Column(name = "start_at")
    private Instant startAt;

    /** The usage percentage at which no new task or step is started; null means no ceiling. */
    @Column(name = "ceiling_percent")
    private Integer ceilingPercent;

    @Column(name = "skip_permissions", nullable = false)
    private boolean skipPermissions;

    @Column(name = "model", length = 20)
    private String model;

    @Column(name = "effort", length = 20)
    private String effort;

    /** While {@code HOLDING}: when the queue looks at usage again. */
    @Column(name = "hold_until")
    private Instant holdUntil;

    @Column(name = "hold_reason", length = 500)
    private String holdReason;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    /** Move to {@code next}, clearing whatever belonged only to the state being left. */
    public void moveTo(RunQueueState next) {
        state = next;
        if (next != RunQueueState.SCHEDULED) {
            startAt = null;
        }
        if (next != RunQueueState.HOLDING) {
            holdUntil = null;
            holdReason = null;
        }
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        return other instanceof RunQueue that && id != null && id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "RunQueue[" + state + "]";
    }
}

File: rekall-model/src/run_queue.rs
//! The one run queue: when it starts, the usage ceiling it stops at, and how the sessions it
//! opens are launched. There is a single row; the queued tasks are run queue items.

use rekall_common::{Id, Instant, RekallError};
use sea_orm::entity::prelude::*;

use crate::constraints::{Phase, Violations};
use crate::enums::RunQueueState;

#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
#[sea_orm(table_name = "run_queue")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: Id,
    pub state: RunQueueState,
    /// When a scheduled queue starts; null for one started at once or not started at all.
    pub start_at: Option<Instant>,
    /// The usage percentage at which no new task or step is started; null means no ceiling.
    pub ceiling_percent: Option<i32>,
    pub skip_permissions: bool,
    pub model: Option<String>,
    pub effort: Option<String>,
    /// While `HOLDING`: when the queue looks at usage again.
    pub hold_until: Option<Instant>,
    pub hold_reason: Option<String>,
    pub created_at: Instant,
    pub updated_at: Instant,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}

impl Model {
    /// `new RunQueue()`: idle, no ceiling, nothing scheduled.
    pub fn new(now: Instant) -> Self {
        Self {
            id: Id::random(),
            state: RunQueueState::Idle,
            start_at: None,
            ceiling_percent: None,
            skip_permissions: false,
            model: None,
            effort: None,
            hold_until: None,
            hold_reason: None,
            created_at: now,
            updated_at: now,
        }
    }

    /// `RunQueue.moveTo`: move to `next`, clearing whatever belonged only to the state being left.
    pub fn move_to(&mut self, next: RunQueueState) {
        self.state = next;
        if next != RunQueueState::Scheduled {
            self.start_at = None;
        }
        if next != RunQueueState::Holding {
            self.hold_until = None;
            self.hold_reason = None;
        }
    }

    pub fn validate(&self) -> Result<(), RekallError> {
        Violations::new("RunQueue", Phase::Update)
            .column("model", self.model.as_deref(), 20)
            .column("effort", self.effort.as_deref(), 20)
            .column("hold_reason", self.hold_reason.as_deref(), 500)
            .finish()
    }
}
